monitor/metrics/checker.go

```diff
@@ -139,31 +140,60 @@ func (mc *Checker) Watch(ctx context.Context, peersF func(context.Context) ([]pe
 // Peers that are not present in the metrics store will return
 // as failed.
 func (mc *Checker) Failed(pid peer.ID) bool {
-	_, _, _, result := mc.failed("ping", pid)
+	_, _, _, result := mc.failed("ping", pid, false)
 	return result
 }
 
 // FailedMetric is the same as Failed but can use any metric type,
 // not just ping.
 func (mc *Checker) FailedMetric(metric string, pid peer.ID) bool {
-	_, _, _, result := mc.failed(metric, pid)
+	_, _, _, result := mc.failed(metric, pid, false)
 	return result
 }
 
 // failed returns all the values involved in making the decision
-// as to whether a peer has failed or not. This mainly for debugging
-// purposes.
-func (mc *Checker) failed(metric string, pid peer.ID) (float64, []float64, float64, bool) {
+// as to whether a peer has failed or not. The debugging parameter
+// enables a more computation heavy path of the function but
+// allows for insight into the return phi value.
+// This should not be used for anything other than testing.
+func (mc *Checker) failed(metric string, pid peer.ID, debugging bool) (float64, []float64, float64, bool) {
 	latest := mc.metrics.PeerLatest(metric, pid)
 	if latest == nil {
 		return 0.0, nil, 0.0, true
 	}
+
+	// the invalidTTL check prevents false-positive results
+	// where multiple metrics closer together skew the distribution
+	// to be less than that of the TTL value of the metrics
+	pmtrs := mc.metrics.PeerMetricAll(metric, pid)
+	var withinTTL bool
+	switch {
+	case len(pmtrs) == 1:
+		if pmtrs[0].Expired() {
+			return 0.0, nil, 0.0, true
+		}
+		return 0.0, nil, 0.0, false
+	case len(pmtrs) >= 2:
+		currMetricExpiry := time.Unix(0, pmtrs[1].Expire)
+		prevMetricReceived := time.Unix(0, pmtrs[0].ReceivedAt)
+		withinTTL = prevMetricReceived.Before(currMetricExpiry)
```

Comment:
ok, re-reading. If there is just one metric and it has not expired -> not-failed

if there is more than one metric, `withinTTL` says if the previous metric (0? isn't 0 the last metric and not the previous) is received before the current metric (1?) expired. In which case -> not failed.

So in order for an alert to trigger, the new metric will have to be received outside the previous metric TTL. However, the distribution may be skewed already after a restart (if two metrics came very close apart vs the general trend), so it will likely not tolerate any issue and send alerts right away. Specially considering that we send metrics at ttl/2 intervals. When it gets a metric outside the TTL of previous it will mean a big deviation. Thus we probably lose the benefits of accrual.

This is a hard problem, I'm not sure when we should alert (or when not: what cases, when the last metric has expired, should not trigger an alert and should just "wait").

Reply:
Note, this affects less when TTLs are low (my problems were with a 5m TTL which obviously gets a big skew in accrual if two metrics are sent a few seconds apart). We the default ping interval is 15 secs.